pkg/storage/registry/dynamic/routingtree/routingtree.go

```diff
@@ -0,0 +1,150 @@
+// Copyright 2018-2023 CERN
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// In applying this license, CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+// Package routingtree contains the routing tree implementation.
+package routingtree
+
+import (
+	"errors"
+	"path"
+	"strings"
+
+	registrypb "github.com/cs3org/go-cs3apis/cs3/storage/registry/v1beta1"
+)
+
+// Route represents a route inside a storage provider.
+type Route struct {
+	Name      string
+	MountID   string
+	MountPath string
+}
+
+// RoutingTree is a tree containing routes.
+type RoutingTree struct {
+	route Route
+	nodes map[string]*RoutingTree
+}
+
+// New returns a new RoutingTree.
+func New(routes map[string]string) *RoutingTree {
+	t := RoutingTree{
+		nodes: map[string]*RoutingTree{},
+		route: Route{},
+	}
+
+	for r, m := range routes {
+		t.addRoute(r, m)
+	}
+
+	return &t
+}
+
+func (t *RoutingTree) addNode(r Route) *RoutingTree {
+	newNode, ok := t.nodes[r.Name]
+	if !ok {
+		newNode = &RoutingTree{
+			route: r,
+			nodes: make(map[string]*RoutingTree),
+		}
+		t.nodes[r.Name] = newNode
+	}
+
+	return newNode
+}
+
+func (t *RoutingTree) addRoute(route, mountID string) {
+	parts := strings.Split(path.Clean(route), "/")
+	current := t
+
+	for i, name := range parts {
+		if name == "" {
+			continue
+		}
+
+		newNode := Route{
+			Name: name,
+		}
+
+		if i == len(parts)-1 {
+			newNode.MountID = mountID
+			newNode.MountPath = route
+		}
+
+		current = current.addNode(newNode)
+	}
+}
+
+func (t *RoutingTree) findRoute(p string) (*RoutingTree, error) {
+	parts := strings.Split(path.Clean(p), "/")
```

Comment:
Use https://github.com/cs3org/reva/blob/d36bcd56dd658a9083495df6c3cae3a129812921/pkg/rhttp/router/router.go#L31 to navigate the path instead of creating a slice with all the tokens

Reply:
Done!